keps/sig-node/0028-20180906-quotas-for-ephemeral-storage.md

```diff
@@ -0,0 +1,731 @@
+  small changes to all volume plugins and their tests, but will in the
+  future allow adding additional data without having to change code
+  other than that which uses the new information.
+  
+#### Testing Strategy
+
+The quota code is by an large not very amendable to unit tests.  While
+there are simple unit tests for parsing the mounts file, and there
+could be tests for parsing the projects and projid files, the real
+work (and risk) involves interactions with the kernel and with
+multiple instances of this code (e. g. in the kubelet and the runtime
+manager, particularly under stress).  It also requires setup in the
+form of a prepared filesystem.  It would be better served by
+appropriate end to end tests.
+
+### Risks and Mitigations
+
+* The SIG raised the possibility of a container being unable to exit
+  should we enforce quotas, and the quota interferes with writing the
+  log.  This can be mitigated by either not applying a quota to the
+  log directory and using the du mechanism, or by applying a separate
+  non-enforcing quota to the log directory.
+
+  As log directories are write-only by the container, and consumption
+  can be limited by other means (as the log is filtered by the
+  runtime), I do not consider the ability to write uncapped to the log
+  to be a serious exposure.
+
+  Note in addition that even without quotas it is possible for writes
+  to fail due to lack of filesystem space, which is effectively (and
+  in some cases operationally) indistinguishable from exceeding quota,
+  so even at present code must be able to handle those situations.
+  
+* Filesystem quotas may impact performance to an unknown degree.
+  Information on that is hard to come by in general, and one of the
+  reasons for using quotas is indeed to improve performance.  If this
+  is a problem in the field, merely turning off quotas (or selectively
+  disabling project quotas) on the filesystem in question will avoid
+  the problem.  Against the possibility that that cannot be done
+  (because project quotas are needed for other purposes), we should
+  provide a way to disable use of quotas altogether via a feature
+  gate.
+  
+  A report <https://blog.pythonanywhere.com/110/> notes that an
+  unclean shutdown on Linux kernel versions between 3.11 and 3.17 can
+  result in a prolonged downtime while quota information is restored.
+  Unfortunately, [the link referenced
+  here](http://oss.sgi.com/pipermail/xfs/2015-March/040879.html) is no
+  longer available.
+
+* Bugs in the quota code could result in a variety of regression
+  behavior.  For example, if a quota is incorrectly applied it could
+  result in ability to write no data at all to the volume.  This could
+  be mitigated by use of non-enforcing quotas.  XFS in particular
+  offers the pqnoenforce mount option that makes all quotas
+  non-enforcing.
+
+  We should offer two feature gates, one to enable quotas at all (on
```

Comment:
A thought: Can we enable quota for monitoring side-by-side with `du`, so that we can compare how stable it is?

Reply:
There's no guarantee that the numbers will match up at any given moment in time.  That's partly because of the issue of deleted files, and partly because a quota reflects a snapshot at a given moment in time while du is not atomic.

The caller could use both mechanisms and compare them; the code calling into the quota for this purpose is isolated.